Project automation: Rewrite actions using JavaScript

noisysocks opened the pull request:
Replaces all of our Docker container actions with a single [JavaScript action](https://help.github.com/en/articles/about-actions#types-of-actions).

This has a few benefits:

- Easier to understand: Gutenberg is primarily a JavaScript application, and we're all primarily JavaScript developers, so it makes sense to use what we know.
- More performant: According to [the docs](https://help.github.com/en/articles/about-actions#types-of-actions), JavaScript actions run faster than Docker container actions.
- Testable: We can use our existing jest setup to write unit tests for our actions.
- Less noise: By replacing 3 actions with 1 action, we'll have 1/3 as many 'checks' appear at the bottom of each PR.

Docs: https://help.github.com/en/categories/automating-your-workflow-with-github-actions

noisysocks committed:
Project automation: Rewrite actions using JavaScript

noisysocks commented on .github/workflows/pull-request-automation.yml:
```diff
@@ -0,0 +1,15 @@
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@master
+      - run: npm install
```
This is necessary so that our action has `@actions/core` and `@actions/github`, but it makes the workflow very very slow 😢

Two alternatives I can think of:

1. Host the action in a different repository and publish the build files. This is [what GitHub recommends](https://github.com/actions/toolkit/blob/master/docs/javascript-action.md). It sucks that we can't use all of the nice infrastructure (e.g. linting, tests) that we already have, though.
2. Change this line so that it only installs the dependencies we know we need: `npm install @actions/core @actions/github`. This would be difficult to maintain, though.

pento replied:
Using `npm ci` instead of `npm install` will cut a bit of time off this (20 seconds testing locally).

talldan replied:
can it install only the's package packages dependencies by cd'ing into the folder running into the command? e.g. `cd packages/pull-request-automation && npm install `.

There'd be a `node_modules` in the local folder, which is a bit different, but I think it's ok as it'll only exist in the context of this job where nobody will be able to see the file system.

noisysocks replied:
> can it install only the package dependencies by cd'ing into the folder running into the command? e.g. cd packages/pull-request-automation && npm install .

Did this in 3c22f92. It's not elegant but let's see how it goes. Thanks for the suggestion @talldan! ❤️

noisysocks commented on .github/workflows/pull-request-automation.yml:
```diff
@@ -0,0 +1,15 @@
+      - run: npm install
+        env:
+          PUPPETEER_SKIP_CHROMIUM_DOWNLOAD: true
+      - run: npm run build:packages
```
I could eliminate this step by not using Babel transformations for the action. Am thinking it's worth it—`import` syntax is the only thing we're using that Node doesn't have by default.

talldan replied:
Agreed, probably not worth keeping it just for `import`.

noisysocks replied:
Removed Babel compilation in 3c22f92 👌

noisysocks commented:
Yay, [it works](https://github.com/WordPress/gutenberg/pull/17080/checks?check_run_id=196706369)! 

I'm pretty happy with where this is at, although I'd like to remove the `npm install` and `npm run build:packages` steps since they're so slow.

What do you think, @talldan?

talldan commented on package.json:
```diff
@@ -64,6 +64,8 @@
 	},
 	"devDependencies": {
+		"@actions/core": "1.0.0",
+		"@actions/github": "1.0.0",
```
Hmm, I wouldn't have thought these dependencies were needed in the root package.json.

swissspidy replied:
According to https://github.com/WordPress/gutenberg/blob/4665d7074dfeae43eac02b1d2e4ad5713832078e/packages/README.md#development-dependencies it's intended that way.

talldan replied:
Thanks for the pointer @swissspidy.

aduth replied:
But these are also listed as `dependencies` in `packages/project-management-automation/package.json` . Surely they don't need to be declared twice? They seem more appropriate as `dependencies` to me than as `devDependencies`, since they're required for the runtime script of the module.

talldan commented on packages/pull-request-automation/src/add-first-time-contributor-label.js:
```diff
@@ -0,0 +1,20 @@
+	} );
+
+	if ( totalCount !== 0 ) {
+		return;
```
Would be cool to have some extra logging output around early returns, as it really helps work out why an action did something/nothing.

noisysocks replied:
Added a tonne of logging in c22a98e0ea1419abcfee2072b9617ef509185064 👌

talldan commented on packages/pull-request-automation/src/assign-fixed-issues.js:
```diff
@@ -0,0 +1,22 @@
+		await octokit.issues.addAssignees( {
+			owner: context.payload.repository.owner.login,
+			repo: context.payload.repository.name,
+			issue_number: +issue,
```
Is the `+` to coerce it into a number?

noisysocks replied:
Yes!

talldan commented on packages/pull-request-automation/src/index.js:
```diff
@@ -0,0 +1,48 @@
+			await assignFixedIssues( context, octokit );
+		} catch ( error ) {
+			setFailed( `assign-fixed-issues: ${ error }` );
+			return;
```
Does this mean the other actions won't run if there's an error with assignFixedIssues?

noisysocks replied:
Good point, that doesn't really make sense. Fixed in c22a98e0ea1419abcfee2072b9617ef509185064 👌

talldan reviewed (approved):
Love this, introduces a really nice scaffolding to add more automation to PRs, and ensures there won't be any extra noise if we do. ❤️ 

I'd be happy to see it merged and have smaller follow up PRs address the comments.

youknowriad commented on packages/pull-request-automation/src/index.js:
```diff
@@ -0,0 +1,48 @@
+		} catch ( error ) {
+			setFailed( `add-milestone: ${ error }` );
+		}
+	}
```
Nit: I feel like there's some redunduncy in this file and this can be probably factorized (composed with some config).

noisysocks replied:
Done in c22a98e0ea1419abcfee2072b9617ef509185064 👌

youknowriad reviewed (commented):
Nice work here 🎉

noisysocks committed:
Project automation: Don't transpile or install all dependencies

noisysocks committed:
Project automation: pull-request-automation -> project-management-automation

noisysocks committed:
Project automation: Add explanatory comment for `npm install` hack

noisysocks committed:
Project automation: Add debug statements

noisysocks committed:
Project automation: Don't use GitHub's debug() function

noisysocks committed:
Project automation: Use `payload` in tasks, not `context`

noisysocks committed:
Project automation: Link to the relevant GitHub documentation

noisysocks commented:
@talldan @youknowriad: Pretty happy with this now! Have moved a few things around so will need another review.

- I renamed the package from `pull-request-automation` to `project-management-automation` in anticipation of future tasks that might do more than just manage pull requests.
- I'm running `npm install` inside the action's directory which makes the action run a lot faster.
- I removed Babel transpilation from the package which makes the action run a lot faster.
- I added lots of logging which should help us work out out what's going wrong/on.
- I fixed a mistake where I was looking at `context.action` instead of `context.payload.action` 🙄

talldan commented on .github/workflows/pull-request-automation.yml:
```diff
@@ -0,0 +1,14 @@
+on: pull_request
+name: Pull request automation
```
Wrong name!

talldan replied:
Also looks like the file name might need to be updated.

noisysocks replied:
The _workflow_ has `on: pull_request` and so is named "Pull request automation". It uses an action called "Project management automation"

talldan replied:
ok ✅

talldan reviewed (approved):
Double approving!
